migration/src/bin/migrate_to.rs

```diff
@@ -0,0 +1,393 @@
+use clap::{builder::PossibleValuesParser, command, Parser, ValueEnum};
+use migration::{
+    sea_orm::{ConnectOptions, Database, DatabaseConnection, EntityTrait, QueryOrder},
+    MigratorTrait,
+};
+use sea_orm_migration::{prelude::*, seaql_migrations};
+use tracing::{info, Level};
+
+#[derive(Copy, Clone, ValueEnum, Debug)]
+enum Direction {
+    Up,
+    Down,
+}
+
+/// Wraps the SeaORM migration library to provide additional features. Lets
+/// you bring migrations up or down to a particular version. Allows dry-run
+/// of migrations.
+#[derive(Parser, Debug)]
+#[command(about, version)]
+struct Args {
+    #[arg(value_enum)]
+    direction: Direction,
+    #[arg(value_parser = available_migrations())]
+    target_version: String,
+    #[arg(short, long)]
+    dry_run: bool,
+    #[arg(short = 'u', long, env = "DATABASE_URL")]
+    database_url: String,
+}
+
+#[async_std::main]
+async fn main() -> Result<(), Error> {
+    let args = Args::parse();
+    tracing_subscriber::fmt()
+        .with_max_level(Level::DEBUG)
+        .init();
+
+    let db = Database::connect(ConnectOptions::new(args.database_url)).await?;
+    migrate_to::<migration::Migrator>(&db, args.dry_run, args.direction, &args.target_version)
+        .await?;
+    Ok(())
+}
+
+async fn migrate_to<M: MigratorTrait>(
+    db: &DatabaseConnection,
+    dry_run: bool,
+    dir: Direction,
+    target: &str,
+) -> Result<(), Error> {
+    let latest_index = match latest_applied_migration(db).await {
+        Ok(m) => migration_index::<M>(&m).ok_or(Error::DBMigrationNotFound)?,
+        Err(_) => -1, // Empty database.
+    };
+    let target_index =
+        migration_index::<M>(target).ok_or(Error::MigrationNotFound(target.to_string()))?;
+    if target_index == latest_index {
+        info!("no action taken, already at desired version");
+        return Ok(());
+    }
+
+    match dir {
+        Direction::Up => match target_index - latest_index {
+            num_migrations if num_migrations > 0 => {
+                info!("executing {num_migrations} migration(s) to {target}");
+                if !dry_run {
+                    M::up(db, Some(num_migrations as u32))
```

Comment:
In the non-dry-run case, do `M::up` and `M::down` log anything about what migrations they have applied?

Relatedly, in the dry-run case, it seems like all we'll log is how many migrations would be applied, but we should log their names/titles as well.

Reply:
The non-dry run case will display queries were applied through its tracing. On further review it's a bit hard to read because the output is out of order due to its asyncness.

> Relatedly, in the dry-run case, it seems like all we'll log is how many migrations would be applied, but we should log their names/titles as well.

:+1:. This should also slightly mitigate the out-of-order log output, the description of what is being executed should always be early in the log output.

Reply:
I was able to have the info log present a slice of migrations being applied, in order. This did regrettably lead to use of `as` to construct ranges, but a cast of u32 to usize should be always safe on any platform we care about.

Reply:
On some further thought, let me move the `u32` cast from above to the call site, then pass around `usize` to eliminate all the `as` statements.